File: cpp_mex/FlightEstimator.cpp
#include "matlab_funcs.h"

#define RTK 1

Matrix9_9 FlightStateTransitionMat(Vector3 accel, Vector3 gyro, Matrix3_3 R_b2i) {
  // Remove angular rates from error-state. Make safety copies of all relevant
  // files into Archive
  Matrix9_9 F = Matrix9_9::Zero();
  F.block<3, 3>(0, 0) = -zetaCross(gyro);
  F.block<3, 3>(3, 6) = Matrix3_3::Identity();
  F.block<3, 3>(6, 0) = -R_b2i * zetaCross(accel);
  return F;
}

Vector13 FlightEstimator(Vector13 x_est, constantsASTRA_t constantsASTRA, Vector15 z, float dT, Matrix9_9 &P, bool new_gps_packet) {
  // M-EKF Implementation
  // Remove bias from IMU
  z.segment<3>(0) = z.segment<3>(0) - x_est.segment<3>(13);
  z.segment<3>(3) = z.segment<3>(3) - x_est.segment<3>(10);
  z.segment<3>(6) = z.segment<3>(6) - x_est.segment<3>(16);

  // Extract quaternion
  Vector9 dx = Vector9::Zero();
  Vector4 q = x_est.segment<4>(0);
  Vector4 qdot = 0.5 * HamiltonianProd(q) * (Vector4() << 0, z.segment<3>(3)).finished();
  x_est.segment<4>(0) = q + qdot * dT;
  q = x_est.segment<4>(0);

  // A-priori quaternion estimate and rotation matrix
  q.normalize();
  Matrix3_3 R_b2i = quatRot(q).transpose();

  // State Transition Matrix
  Matrix9_9 F = FlightStateTransitionMat(z.segment<3>(0), z.segment<3>(3), R_b2i);

  // Propagate rest of state using IMU
  x_est.segment<3>(7) = x_est.segment<3>(7) + (R_b2i * z.segment<3>(0) - (Vector3() << 0, 0, constantsASTRA.g).finished()) * dT;
  x_est.segment<3>(4) = x_est.segment<3>(4) + x_est.segment<3>(7) * dT;

  // Discrete STM
  Matrix9_9 Phi = matrixExpPade6<Matrix9_9>(F * dT);

  // Extract Matrices
  Matrix9_9 Q = constantsASTRA.Q.block<9, 9>(0, 0);

  // Process Noise Covariance and a-priori propagation step
  Q = 0.5 * Q;
  P = Phi * P * Phi.transpose() + Q;

  // GPS Update
  if (new_gps_packet) {
    // Measurement matrix
    Matrix6_9 H = Matrix6_9::Zero();
    H.block<3, 3>(0, 3) = Matrix3_3::Identity();
    H.block<3, 3>(3, 6) = Matrix3_3::Identity();

    // Measurement Covariance Matrix
    float gps_pos_covar = 1 * RTK + 100 * (1 - RTK);
    float gps_vel_covar = gps_pos_covar * 1;
    Matrix6_6 R = (Vector6() << pow(gps_pos_covar, 2) * Vector3::Ones(), pow(gps_vel_covar, 2) * Vector3::Ones()).finished().asDiagonal();

    // A priori covariance and Kalman gain
    Matrix9_6 L = P * H.transpose() * (H * P * H.transpose() + R).inverse();

    // Predicted measurements
    Vector6 z_hat = (Vector6() << x_est.segment<3>(4), x_est.segment<3>(7)).finished();

    // Update Error State
    Matrix9_9 ILH = (Matrix9_9::Identity() - L * H);
    P = ILH * P * ILH.transpose() + L * R * L.transpose();
    Vector6 residual = (z.segment<6>(9) - z_hat);
    Vector9 inn = L * residual;
    dx = dx + inn;
  }

  // Update full-state estimates
  Vector4 dq = (Vector4() << 1, dx.segment<3>(0) / 2).finished();
  dq.normalize();

  Vector4 q_nom = HamiltonianProd(q) * dq;
  q_nom.normalize();
  x_est.segment<4>(0) = q_nom;
  x_est.segment<6>(4) = x_est.segment<6>(4) + dx.segment<6>(3);
  return x_est;
}

File: Filtering/Kalman FIlter/FlightEstimator.m
function [x_est, lastP] = FlightEstimator(x_est,constantsASTRA,z,dT,P0)
%% M-EKF Implementation
% Remove bias from IMU
z(1:3) = z(1:3) - x_est(14:16);
z(4:6) = z(4:6) - x_est(11:13);
z(7:9) = z(7:9) - x_est(17:19);

% Extract quaternion
dx = zeros(9,1);
q = x_est(1:4);
qdot = 0.5 * HamiltonianProd(q) * [0; z(4:6)]; 
x_est(1:4) = q + qdot * dT;
q = x_est(1:4);

% A-priori quaternion estimate and rotation matrix
q = q / norm(q);
R_b2i = quatRot(q)';

% GPS velocity correction
rGPS = [0 0 0.31]';
z(13:15) = z(13:15) - R_b2i * cross(z(4:6), rGPS);

% Process Covariance Matrix
persistent P lastZ 
if isempty(P)
    P = P0;  
    lastZ = zeros(15,1);
end

% State Transition Matrix
F = StateTransitionMat(z(1:3), z(4:6), R_b2i, 0);

% Propagate rest of state using IMU
x_est(8:10) = x_est(8:10) + (R_b2i * z(1:3) - [0; 0; constantsASTRA.g]) * dT;
x_est(5:7) = x_est(5:7) + x_est(8:10) * dT;

% Discrete STM
Phi = expm(F * dT);

% Extract Matrices
Q = constantsASTRA.Q(1:9,1:9);

% Process Noise Covariance and a-priori propagation step
Q = 0.5 * Q;
P = Phi * P * Phi' + Q;
RTK = 1;

%% GPS Update
if sum(lastZ(10:15) - z(10:15)) ~=0

    % Measurement matrix
    H = zeros(6,9);
    H(1:3, 4:6) = eye(3);
    H(4:6, 7:9) = eye(3);

    % Measurement Covariance Matrix
    gps_pos_covar = 0.2 * RTK + 10 * (1 - RTK);
    gps_vel_covar = 0.05; %gps_pos_covar * 5;
    R = diag([gps_pos_covar^2 * ones(3,1); gps_vel_covar^2 * ones(3,1)]);

    % A priori covariance and Kalman gain
    L = P * H' / (H * P * H' + R);

    % Predicted measurements 
    z_hat = [x_est(5:7);
             x_est(8:10)];

    % Update Error State
    ILH = (eye(9) - L * H);
    P = ILH * P * ILH' + L * R * L';
    residual = (z(10:15) - z_hat);
    inn = L * residual;
    dx = dx + inn;
end

% Update full-state estimates
dq = [1; dx(1:3) / 2];
dq = dq / norm(dq);

q_nom = quatmultiply(q', dq');
q_nom = q_nom / norm(q_nom); 
x_est(1:4) = q_nom';
x_est(5:10) = x_est(5:10) + dx(4:9);
lastZ = z;
lastP = P;
end